examples/restriction/non_topologically_sorted_functions/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+#![feature(rustc_private)]
+#![warn(unused_extern_crates)]
+
+extern crate rustc_hir;
+extern crate rustc_span;
+
+use rustc_hir::def::Res;
+use rustc_hir::def_id::LocalDefId;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_hir::{BodyId, Expr, ExprKind, HirId, Item, ItemKind, Mod};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_span::Span;
+use std::collections::{HashMap, HashSet};
+
+dylint_linting::declare_late_lint! {
+    /// ### What it does
+    ///
+    /// It enforces a certain relative order among functions defined within a module.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Without a certain order it's really bad to navigate through the modules.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn bar() { }
+    ///
+    /// fn foo() {
+    ///     bar();
+    /// }
+    /// ```
+    ///
+    /// Use instead:
+    ///
+    /// ```rust
+    /// fn foo() {
+    ///     bar();
+    /// }
+    ///
+    /// fn bar() { }
+    /// ```
+    pub NON_TOPOLOGICALLY_SORTED_FUNCTIONS,
+    Warn,
+    "Enforce callers before callees and consistent order of callees (module-local functions)"
+}
+
+struct Finder<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    local_defs: &'a HashMap<LocalDefId, Span>,
+    seen: HashSet<LocalDefId>,
+    order: Vec<LocalDefId>,
+}
+
+impl<'tcx> Visitor<'tcx> for Finder<'_, 'tcx> {
+    fn visit_expr(&mut self, ex: &'tcx Expr<'tcx>) {
+        if let ExprKind::Call(callee, _args) = &ex.kind
+            && let ExprKind::Path(ref qpath) = callee.kind
+        {
+            let res = self.cx.qpath_res(qpath, callee.hir_id);
+            if let Res::Def(_, def_id) = res
+                && let Some(local_def_id) = def_id.as_local()
+                && self.local_defs.contains_key(&local_def_id)
+                && !self.seen.contains(&local_def_id)
```

Comment:
Could these be combined into a single if-let?

Reply:
Something like this:
```suggestion
        if let ExprKind::Call(callee, _args) = &ex.kind
            && let ExprKind::Path(ref qpath) = callee.kind
            && let res = self.cx.qpath_res(qpath, callee.hir_id);
            && let Res::Def(_, def_id) = res
            && let Some(local_def_id) = def_id.as_local()
            && self.local_defs.contains_key(&local_def_id)
            && !self.seen.contains(&local_def_id)
```

Reply:
did this